### ISSUE TITLE
Support salt 3000.2

### ISSUE THREAD

**Issue by TeddyAndrieux** [scality/metalk8s#650]
Currently salt 2019.2.0 is not supported as we use yaml information directly. 
So unicode instead of string

https://docs.saltstack.com/en/latest/topics/releases/2019.2.0.html#non-backward-compatible-change-to-yaml-renderer

```
bootstrap: [ERROR   ] Command '[u'systemd-run', u'--scope', u'yum', u'-y', u'install', u"u'util-linux'", u"u'iptables'", u"u'iproute'"]' failed with return code: 1
    bootstrap: [ERROR   ] stdout: Running scope as unit run-396.scope.
    bootstrap: Loaded plugins: fastestmirror
    bootstrap: Loading mirror speeds from cached hostfile
    bootstrap:  * base: distrib-coffee.ipsl.jussieu.fr
    bootstrap:  * epel: mirror.ibcp.fr
    bootstrap:  * extras: centos.quelquesmots.fr
    bootstrap:  * updates: centos.mirrors.proxad.net
    bootstrap: No package u'util-linux' available.
    bootstrap: No package u'iptables' available.
    bootstrap: No package u'iproute' available.
    bootstrap: Error: Nothing to do
    bootstrap: [ERROR   ] retcode: 1
    bootstrap: [ERROR   ] Error occurred installing package(s). Additional info follows:
    bootstrap: 
    bootstrap: errors:
    bootstrap:     - Running scope as unit run-396.scope.
    bootstrap:       Loaded plugins: fastestmirror
    bootstrap:       Loading mirror speeds from cached hostfile
    bootstrap:        * base: distrib-coffee.ipsl.jussieu.fr
    bootstrap:        * epel: mirror.ibcp.fr
    bootstrap:        * extras: centos.quelquesmots.fr
    bootstrap:        * updates: centos.mirrors.proxad.net
    bootstrap:       No package u'util-linux' available.
    bootstrap:       No package u'iptables' available.
    bootstrap:       No package u'iproute' available.
    bootstrap:       Error: Nothing to do
    bootstrap: [ERROR   ] Command '[u'systemd-run', u'--scope', u'yum', u'-y', u'install', u"u'ethtool'", u"u'socat'", u"u'ebtables'", u"u'coreutils'", u"u'iproute'"]' failed with return code: 1
    bootstrap: [ERROR   ] stdout: Running scope as unit run-470.scope.
    bootstrap: Loaded plugins: fastestmirror
    bootstrap: Loading mirror speeds from cached hostfile
    bootstrap:  * base: distrib-coffee.ipsl.jussieu.fr
    bootstrap:  * epel: mirror.ibcp.fr
    bootstrap:  * extras: centos.quelquesmots.fr
    bootstrap:  * updates: centos.mirrors.proxad.net
    bootstrap: No package u'ethtool' available.
    bootstrap: No package u'socat' available.
    bootstrap: No package u'ebtables' available.
    bootstrap: No package u'coreutils' available.
    bootstrap: No package u'iproute' available.
    bootstrap: Error: Nothing to do
    bootstrap: [ERROR   ] retcode: 1
    bootstrap: [ERROR   ] Error occurred installing package(s). Additional info follows:
    bootstrap: 
    bootstrap: errors:
    bootstrap:     - Running scope as unit run-470.scope.
    bootstrap:       Loaded plugins: fastestmirror
    bootstrap:       Loading mirror speeds from cached hostfile
    bootstrap:        * base: distrib-coffee.ipsl.jussieu.fr
    bootstrap:        * epel: mirror.ibcp.fr
    bootstrap:        * extras: centos.quelquesmots.fr
    bootstrap:        * updates: centos.mirrors.proxad.net
    bootstrap:       No package u'ethtool' available.
    bootstrap:       No package u'socat' available.
    bootstrap:       No package u'ebtables' available.
    bootstrap:       No package u'coreutils' available.
    bootstrap:       No package u'iproute' available.
    bootstrap:       Error: Nothing to do
```

**Comment by NicolasT** [comment]
If we do this, I'd suggest we don't just "support" 2019.2.0, because in the end we (currently) own the whole stack, but go all the way and deploy those versions of master and minion as well.

**Comment by gdemonet** [comment]
Agreed, I'll also go further and deploy Python 3 versions as well (which in turn implies packaging of some dependencies, e.g. `python34-kubernetes`)

**Comment by gdemonet** [comment]
**Update**: using Salt 2019.2.0 is currently blocked due to https://github.com/saltstack/salt/issues/51869

It seems using Python 3 does not raise this bug, however it implies large amounts of packaging for dependencies missing in EPEL/Salt repository (in particular for the `python-kubernetes` client).

Thus, we will wait for a fix upstream and let the minions run in Python2, until #626 and #691 move all packaging needs to the Salt-master container only.

**Comment by NicolasT** [comment]
#626 and #691 are now closed.

**Comment by NicolasT** [comment]
Would it make sense to monkey-patch around this issue as hinted at in https://github.com/saltstack/salt/issues/51869#issuecomment-489219070, or alternatively copy a 'fixed' version of the module/procedure in an extmod we ship?

I'd like to move forward with 2019.2 without waiting for 2019.2.1 being released, if possible...

**Comment by gdemonet** [comment]
I guess we can use the monkey-patch, since we already have our own `metalk8s_kubernetes` state module, why not one more?

**Comment by NicolasT** [comment]
Exactly.

At the same time, whenever we  copy some code from Salt `master` or some later release, we should clearly tag this somehow in our code, such that whenever we move to a new Salt release, we can easily find those copies and remove them when applicable.

I failed to do so in the past myself, will do so in two places at least. Let's discuss to find a proper common format for this.

**Comment by shqnayaz** [comment]
HI Nicolas/Team, 
I am trying to upgrade the ring from 6.4.5.10 to 7. in my local Lab
and when i run the upgrade script it fails

# ./upgrade
[2019-08-01 11:51:47+01:00] <salt> Update modules ... OK
[2019-08-01 11:51:49+01:00] <repo> Start repository HTTP server ... ERR: unable to apply Salt state scality.bootstrap.installed

where as all my minions are up 
any suggestions please ?

**Comment by NicolasT** [comment]
Hello @shqnayaz,

Thanks for reaching out.

Your question relates to the installation/upgrade of Scality RING, which is a product not related to MetalK8s, the open-source project backed by Scality developed in this repository.

Please refer to the Scality RING documentation and the support channels available to you as a Scality customer so we can help you further.

Thank you,

Nicolas

**Comment by NicolasT** [comment]
The blocker is now fixed in upstream Salt releases v2019.2.1 and v2019.2.2.

**Comment by TeddyAndrieux** [comment]
Some changes need to be done on MetalK8s side to support salt 3000.2

- Cast to string when checking available version in our custom package manager module (as version could be an int or a float `3000.2` is a float by default)
```
diff --git a/salt/_modules/metalk8s_package_manager_yum.py b/salt/_modules/metalk8s_package_manager_yum.py
index 062e6346..386ced1b 100644
--- a/salt/_modules/metalk8s_package_manager_yum.py
+++ b/salt/_modules/metalk8s_package_manager_yum.py
@@ -116,7 +116,7 @@ def list_pkg_dependents(
     if not version:
         return all_pkgs

-    if pkgs_info and versions_dict[name] != version:
+    if pkgs_info and str(versions_dict[name]) != str(version):
         log.error(
             'Trying to list dependents for "%s" with version "%s", '
             'while version configured is "%s"',
```

- For each and every certificate advertise apply `tojson` filter to avoid unicode certs
```
diff --git a/salt/metalk8s/kubernetes/ca/etcd/advertised.sls b/salt/metalk8s/kubernetes/ca/etcd/advertised.sls
index a10e9702..40fd4ccf 100644
--- a/salt/metalk8s/kubernetes/ca/etcd/advertised.sls
+++ b/salt/metalk8s/kubernetes/ca/etcd/advertised.sls
@@ -13,7 +13,7 @@ Ensure etcd CA cert is present:
     - mode: 644
     - makedirs: True
     - dir_mode: 755
-    - contents: {{ etcd_ca.splitlines() }}
+    - contents: {{ etcd_ca.splitlines() | tojson }}

 {%- else %}
```
